File: Práctica12/PatrondeComportamiento.java
import java.util.*;

interface Observador {
    void actualizar(String mensaje);
}

class ObservadorA implements Observador {
    private String nombre;

    public ObservadorA(String nombre) {
        this.nombre = nombre;
    }

    public void actualizar(String mensaje) {
        System.out.println(nombre + " recibió: " + mensaje);
    }
}

class Sujeto {
    private List<Observador> observadores = new ArrayList<>();

    public void agregar(Observador o) {
        observadores.add(o);
    }

    public void notificar(String mensaje) {
        for (Observador o : observadores) {
            o.actualizar(mensaje);
        }
    }
}

class Main {
    public static void main(String[] args) {
        Sujeto s = new Sujeto();
        s.agregar(new ObservadorA("Cliente 1"));
        s.agregar(new ObservadorA("Cliente 2"));
        s.notificar("Nuevo producto disponible");
    }
}

File: Práctica12/PatrondeComportamiento.py
class Observador:
    def actualizar(self, mensaje):
        pass

class ObservadorA(Observador):
    def __init__(self, nombre):
        self.nombre = nombre

    def actualizar(self, mensaje):
        print(f"{self.nombre} recibió: {mensaje}")

class Sujeto:
    def __init__(self):
        self.observadores = []

    def agregar(self, observador):
        self.observadores.append(observador)

    def notificar(self, mensaje):
        for observador in self.observadores:
            observador.actualizar(mensaje)

if __name__ == "__main__":
    sujeto = Sujeto()
    sujeto.agregar(ObservadorA("Cliente 1"))
    sujeto.agregar(ObservadorA("Cliente 2"))
    sujeto.notificar("Nuevo producto disponible")